src/frequenz/dispatch/actor.py

```diff
@@ -0,0 +1,279 @@
+            )
+
+    async def _schedule_task(self, dispatch: Dispatch) -> None:
+        """Wait for a dispatch to become ready.
+
+        Waits for the dispatches next run and then notifies that it is ready.
+
+        Args:
+            dispatch: The dispatch to schedule.
+        """
+
+        def next_run_info() -> tuple[datetime, datetime] | None:
+            now = datetime.now(tz=timezone.utc)
+            next_run = self.calculate_next_run(dispatch, now)
+
+            if next_run is None:
+                return None
+
+            return now, next_run
+
+        while pair := next_run_info():
+            now, next_time = pair
+
+            if next_time - now > _MAX_AHEAD_SCHEDULE:
+                await asyncio.sleep(_MAX_AHEAD_SCHEDULE.total_seconds())
+                continue
+
+            _logger.info("Dispatch %s scheduled for %s", dispatch.id, next_time)
+            await asyncio.sleep((next_time - now).total_seconds())
+
+            _logger.info("Dispatch ready: %s", dispatch)
+            await self._ready_dispatch_sender.send(dispatch)
+
+        _logger.info("Dispatch finished: %s", dispatch)
+        self._scheduled.pop(dispatch.id)
+
+    @classmethod
+    def calculate_next_run(cls, dispatch: Dispatch, _from: datetime) -> datetime | None:
+        """Calculate the next run of a dispatch.
+
+        Args:
+            dispatch: The dispatch to calculate the next run for.
+            _from: The time to calculate the next run from.
+
+        Returns:
+            The next run of the dispatch or None if the dispatch is finished.
+        """
+        if (
+            not dispatch.recurrence.frequency
+            or dispatch.recurrence.frequency == Frequency.UNSPECIFIED
+        ):
+            if _from > dispatch.start_time:
+                return None
+            return dispatch.start_time
+
+        # Make sure no weekday is UNSPECIFIED
+        if Weekday.UNSPECIFIED in dispatch.recurrence.byweekdays:
+            _logger.warning(
+                "Dispatch %s has UNSPECIFIED weekday, ignoring...", dispatch.id
+            )
+            return None
+
+        count, until = (None, None)
+        if end := dispatch.recurrence.end_criteria:
+            count = end.count
+            until = end.until
+
+        next_run = rrule.rrule(
+            freq=_RRULE_FREQ_MAP[dispatch.recurrence.frequency],
+            dtstart=dispatch.start_time,
+            count=count,
+            until=until,
+            byminute=dispatch.recurrence.byminutes,
+            byhour=dispatch.recurrence.byhours,
+            byweekday=[
+                _RRULE_WEEKDAY_MAP[weekday]
+                for weekday in dispatch.recurrence.byweekdays
+            ],
+            bymonthday=dispatch.recurrence.bymonthdays,
+            bymonth=dispatch.recurrence.bymonths,
+            interval=dispatch.recurrence.interval,
+        )
+
+        return cast(datetime | None, next_run.after(_from, inc=True))
```

Comment:
Is this `cast` needed because `dateutil` lacks type info or what?

Reply:
This one is too small to create an issue. So also please comment here.

Reply:
Yah, I was getting `src/frequenz/dispatch/actor.py:279: error: Returning Any from function declared to return "datetime | None"  [no-any-return]`

Reply:
OK, it is probably best to add a comment when using `cast`, so readers don't have to guess in the future.